### PR TITLE
Move AON, AONPT counter and QMSI RTC to unified interface

### PR DESCRIPTION
**Pull request by SavinayDharmappa**
This is a series of patch set which moves aon, aonpt and rtc_qmsi driver api to proposed unifying interface in PR #8340 . It also changes existing test applications to work with unified interface.

### CODE REVIEW

File: drivers/counter/counter_qmsi_aonpt.c [hunk 1]
```diff
@@ -108,8 +109,7 @@ static u32_t aon_timer_qmsi_read(struct device *dev)
 }
 
 static int aon_timer_qmsi_set_alarm(struct device *dev,
```
Comment (nordic-krch) [inline drivers/counter/counter_qmsi_aonpt.c]: legacy set_alarm function has been changed to set_wrap_alarm so this function should actually implement interface

```
typedef int (*counter_api_set_wrap_alarm)(struct device *dev,
					  u32_t ticks,
					  counter_wrap_callback_t callback,
					  void *user_data);

set_alarm should be used if device has more compare channels and alarms are set within counter period (which is set by set_wrap_alarm).
```

Comment (nordic-krch) [inline drivers/counter/counter_qmsi_aonpt.c]: legacy set_alarm function has been changed to set_wrap_alarm so this function should actually implement interface

```
typedef int (*counter_api_set_wrap_alarm)(struct device *dev,
					  u32_t ticks,
					  counter_wrap_callback_t callback,
					  void *user_data);

set_alarm should be used if device has more compare channels and alarms are set within counter period (which is set by set_wrap_alarm).
```

File: drivers/counter/counter_qmsi_aonpt.c [hunk 2]
```diff
@@ -17,9 +17,10 @@
 #include "qm_aon_counters.h"
 #include "qm_isr.h"
 
+
 static void aonpt_int_callback(void *user_data);
 
-static counter_callback_t user_cb;
+static counter_alarm_callback_t user_cb;
```
Comment (nordic-krch) [inline drivers/counter/counter_qmsi_aonpt.c]: it should be counter_wrap_callback_t

**Review by nordic-krch — CHANGES_REQUESTED** [review]

**Comment by nashif** [comment]
@SavinayDharmappa can you please address the comments?

File: drivers/counter/counter_qmsi_aonpt.c (hunk 1 above)
Comment (SavinayDharmappa) [reply drivers/counter/counter_qmsi_aonpt.c]: ok . I change it.

Comment (SavinayDharmappa) [reply drivers/counter/counter_qmsi_aonpt.c]: ok . I will do it.

File: drivers/counter/counter_qmsi_aonpt.c (hunk 2 above)
Comment (SavinayDharmappa) [reply drivers/counter/counter_qmsi_aonpt.c]: ok

**Comment by SavinayDharmappa** [comment]
@nordic-krch  I have addressed ur review comments.

**Comment by SavinayDharmappa** [comment]
sanity is failing as i haven't included counter.h

File: drivers/counter/counter_qmsi_aonpt.c [hunk 3]
```diff
@@ -108,8 +109,9 @@ static u32_t aon_timer_qmsi_read(struct device *dev)
 }
 
 static int aon_timer_qmsi_set_alarm(struct device *dev,
-				    counter_callback_t callback,
-				    u32_t count, void *user_data)
+				    u32_t ticks,
+				    counter_wrap_callback_t callback,
+				    void *user_data)
```
Comment (nordic-krch) [inline drivers/counter/counter_qmsi_aonpt.c]: note that new api does not require to set wrap only when counter is already started. So `if (QM_AONC[QM_AONC_0]->aonpt_cfg == 0)` can be removed.  Basically, it should be possible to setup the counter and then start it.

**Review by nordic-krch — DISMISSED** [review]

File: drivers/counter/counter_qmsi_aonpt.c (hunk 3 above)
Comment (SavinayDharmappa) [reply drivers/counter/counter_qmsi_aonpt.c]: done

**Comment by nordic-krch** [comment]
please fix conflicts

**Comment by nashif** [comment]
Please rebase against topic-counter branch and resubmit. Remove already merged patches.

**Comment by nordic-krch** [comment]
please rebase on topic-counters

**Comment by nashif** [comment]
rebased...

**Comment by carlescufi** [comment]
@nordic-krch this is ready for re-review

**Comment by SavinayDharmappa** [comment]
@nashif do I need to rebase it again and push it

**Comment by nashif** [comment]
> @nashif do I need to rebase it again and push it

no

**Review by nashif — CHANGES_REQUESTED** [review]

File: drivers/counter/counter_rtc_qmsi.c [hunk 4]
```diff
@@ -8,6 +8,7 @@
 
 #include <device.h>
 #include <drivers/ioapic.h>
+#include <counter.h>
 #include <init.h>
 #include <kernel.h>
 #include <rtc.h>
```
[the remark this answers is not in the document]
Comment (SavinayDharmappa) [reply drivers/counter/counter_rtc_qmsi.c]: I will remove it

File: drivers/counter/Kconfig [hunk 5]
```diff
@@ -20,6 +20,13 @@ config COUNTER_LOG_LEVEL
 	help
 	  Counter logging level.
 
+config RTC_0_NAME
```
[the remark this answers is not in the document]
Comment (SavinayDharmappa) [reply drivers/counter/Kconfig]: done

File: samples/drivers/rtc/src/main.c [hunk 6]
```diff
@@ -7,35 +7,32 @@
 #include <zephyr.h>
 
 #include <device.h>
+#include <counter.h>
 #include <rtc.h>
```
[the remark this answers is not in the document]
Comment (SavinayDharmappa) [reply samples/drivers/rtc/src/main.c]: will remove it

**Comment by SavinayDharmappa** [comment]
@nordic-krch  I am getting  "warning: ‘counter_callback_t’ is deprecated" even though i am not using counter_callback_t in driver. Can you review my patch

**Review by dbkinder — COMMENTED** [review]
no doc changes for me to review.

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: soc/x86/intel_quark/quark_se/Kconfig.defconfig.series [hunk 7]
```diff
@@ -148,11 +150,6 @@ config WDT_QMSI
 	default y
 endif # WATCHDOG
 
-if RTC
```
[the remark this answers is not in the document]
Comment (SavinayDharmappa) [reply soc/x86/intel_quark/quark_se/Kconfig.defconfig.series]: will change it.

File: soc/x86/intel_quark/quark_d2000/Kconfig.defconfig.series [hunk 8]
```diff
@@ -56,16 +56,6 @@ config WDT_QMSI
 	default y
 endif # WATCHDOG
 
-if RTC
```
[the remark this answers is not in the document]
Comment (SavinayDharmappa) [reply soc/x86/intel_quark/quark_d2000/Kconfig.defconfig.series]: will change it.

File: tests/drivers/aon_counter/aon_api/src/test_aon_periodic_timer.c [hunk 9]
```diff
@@ -68,16 +68,18 @@ static int test_timer(void)
 	TC_PRINT("Always-on timer started\n");
 
 	/* verify counter_set_alarm() */
-	if (counter_set_alarm(aon_timer, aon_timer_callback,
-			      ALARM_CNT, (void *)&dummy_data)) {
+	if (counter_set_wrap(aon_timer, ALARM_CNT,
```
[the remark this answers is not in the document]
Comment (SavinayDharmappa) [reply tests/drivers/aon_counter/aon_api/src/test_aon_periodic_timer.c]: will work towards it.

**Review by nashif — APPROVED** [review]

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**